src/profile-logic/flame-graph.js

```diff
@@ -39,29 +44,135 @@ type Stack = Array<{
+ * an offset array used to index into it.
+ */
+export function getRootsAndChildren(
+  thread: Thread,
+  callNodeTable: CallNodeTable,
+  callNodeChildCount: Uint32Array,
+  totalTime: Float32Array
+) {
+  const roots = [];
+  /* All children for all callnodes will be stored in a large, flat
+   * array called, simply, `array`. In it, siblings are always next to
+   * each other in contiguous slices. To find the children for a
+   * particular callnode, we use a pair of values from `offsets` to
+   * find the right slice. */
+  const array = new Uint32Array(callNodeTable.length);
+  const offsets = new Uint32Array(callNodeTable.length + 1);
+
+  /* For performance reasons the array is of type Uint32Array. This
+   * means we cannot use values such as `undefined` or `null` to
+   * indicate uninitialized values, as we build up the array. But
+   * since `callNodeTable` is ordered is such a way that a given
+   * callnode index always comes _after_ its parent callnode index, we
+   * know that callnode index zero never can be a child. It is always
+   * a root. (Not counting the special -1 root, but we don't need it
+   * here). Hence, we are free to use the value 0 in the children
+   * array to mark elements as not initialized, since 0 is never a
+   * valid child. Since the default values of Uint32Array is 0, we
+   * conveniently get an array where all its values are uninitialized
+   * from start. */
+
+  let callNodeIndex = 0;
+  let ptr = 0;
+  for (; callNodeIndex < callNodeTable.length; callNodeIndex++) {
+    offsets[callNodeIndex] = ptr;
+    ptr += callNodeChildCount[callNodeIndex];
+
+    if (totalTime[callNodeIndex] === 0) {
+      continue;
+    }
+
+    const parent = callNodeTable.prefix[callNodeIndex];
+    if (parent === -1) {
+      roots.push(callNodeIndex);
+      continue;
+    }
+
+    const funcName = thread.stringTable.getString(
+      thread.funcTable.name[callNodeTable.func[callNodeIndex]]
+    );
+
+    /* From the parent, we can now know the slice allotted for all
+     * its children. */
+    const start = offsets[parent];
+    const end = offsets[parent] + callNodeChildCount[parent] - 1;
+
+    /* Find the place in `array` where this callnode should be
+     * inserted, swapping elements in the array as we go
+     * along. Continue as long as this callnode's function name is
+     * lexically greater than the function names of the callnodes
+     * already placed in the array. This ensures that all slices have
+     * children in descending order. Any callnode indices equal to 0
+     * means that they are uninitialized, so just breeze through
+     * them. When we stop, when have found the right position to
+     * insert our callnode.
+     *
+     * This effectively is an insertion sort, which is O(n^2), but
+     * since n is typically small (the number of children of a given
+     * callnode), it should be just fine.
+     */
+    let i = start;
+    while (
```

Comment:
This while check is a bit hard to read, not sure on the best way to fix it, but if is something comes to mind it could make future readers have an easier time.